helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -87,19 +90,30 @@ public Object deserialize(byte[] data) throws ZkMarshallingError {
 
     _zkSerializer = new ZNRecordJacksonSerializer();
     _bucketSize = bucketSize;
+    _numVersions = numVersions;
   }
 
   /**
    * Constructor that uses a default bucket size.
    * @param zkAddr
    */
   public ZkBucketDataAccessor(String zkAddr) {
-    this(zkAddr, DEFAULT_BUCKET_SIZE);
+    this(zkAddr, DEFAULT_BUCKET_SIZE, DEFAULT_NUM_VERSIONS);
   }
 
   @Override
   public <T extends HelixProperty> boolean compressedBucketWrite(String path, T value)
       throws IOException {
+    lock(path);
```

Comment:
@i3wangyi Can you remove this duplicate comment?
It is true that you don't need to establish multiple ZkClient connections for the use case you're referring to. However, this is a data accessor for large writes for more general use. That is why we consider the possibility of multiple writes and reads. This was the original design; the data accessor is not singly tied to the WAGED rebalancer. 

For our WAGED rebalancer use case, there will be a component called AssignmentMetadataStore that maintains only one instance of BucketDataAccessor that will be used across pipeline runs.

Reply:
@i3wangyi https://github.com/apache/helix/issues/444
Here is the issue for the aforementioned component. This will create and maintain one instance of `BucketDataAccessor`. Once we have `BucketDataAccessor` (which is more of a generic component that could be used in other places), I will implement `AssignmentMetadataStore`.

Reply:
I was thinking about this point, for the genericity of the accessor, I think a ZK backed lock is necessary. Eventually, we will need to evaluate the extra cost, of course.

One minor thing may help with the performance a little bit, how about applying the lock after bucket calculation?

Reply:
@jiajunwang @narendly 
It starts to make sense when the implementation of lock will be used in a distributed environment where instances other than helix controller itself will coordinate and write in order. Think about this use case then becomes actually a distributed lock, is using distributed lock really needed for the use case? Will we ever expose the endpoint to other parties?

My in-memory lock works in a single controller environment where the critical path write operations are only possible for one instance. Then singleton pattern + key based lock will work just fine, it'll be more performant since this implementation relies on ZK, which at the end of the day, is an I/O operation/network request, etc.

Does this understanding make sense? Have we cleared about our use cases?

Reply:
We've synced up offline. We do have such a use case where a custom tool or other components will read the same path and introduce a race-condition situation. In other words, the use case is in a distributed environment. Given the lack of a re-usable distributed lock module, the current implementation is a quick and light-weight solution to get things working.